Docs/csp guide (#947)
ionic-team/stencil-site: src/docs/guides/csp-nonce.md

---
title: Content Security Policy Nonces
description: How to leverage CSP nonces in Stencil projects.
url: /docs/csp-nonce
contributors:
  - tanner-reits
---

# Using Content Security Policy Nonces in Stencil

[Content Security Policies (CSPs)](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP) can help protect an application from Cross-Site Scripting (XSS)
attacks by adding a security layer to help prevent unauthorized code from running in the browser.

> In the world of Stencil applications, if your app leverages a restrictive CSP (i.e. anything other than 'unsafe-inline') and components without a Shadow
> DOM, you'll likely run into an error in the browser console stating the certain styles or scripts were not able to run because they violate the
> effective CSP.

Reviewer: ```suggestion
An application that is served with a restrictive CSP other than `'unsafe-inline'` & contains web components without a Shadow DOM will likely run into errors on load. This is often first detected in the browser's console, which reports an error stating that certain styles or scripts violate the effective CSP.
```

article walks through the process using Nginx and Webpack. Obviously, these resources don't account for the Stencil specifics, but we'll try to make
those parts clear as we move on.

Per the [MDN Guide on nonces](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/nonce#generating_values), a nonce should be "a random base64-encoded string of at least 128 bits of data from a cryptographically secure random number generator".

### Output Targets

Using nonces may differ slightly between output targets, so please be sure to use the correct pattern based on the context in which your
Stencil components are consumed.

#### Dist

Consuming a `nonce` in the `dist` output target is easy using the provided `setNonce` helper function. This function is exported from the index
file of the output target's designated output directory.

This function simply accepts the `nonce` string value that you want set for every `style` and `script` tag.

This is an example of consuming the `dist` output in an Angular app's entrypoint:

```ts
// main.ts

import { defineCustomElements, setNonce } from 'my-lib/loader';

// Will set the `nonce` attribute for all scripts/style tags

> // i.e will run styleTag.setAttribute('nonce', 'r4nd0m')

Reviewer: ```suggestion
// i.e. will run styleTag.setAttribute('nonce', 'r4nd0m')
```

from the index file of the output target's designated output directory.

This function simply accepts the `nonce` string value that you want set for every `style` and `script` tag.

This is an example of consuming the `dist-custom-elements` output in an Angular app's entrypoint:

```ts
// main.ts

import { defineCustomElements, setNonce } from 'my-lib/dist/components';
// We'll assume we have `autoDefineCustomElements` enabled
import 'my-lib/dist/components/my-component';

// Will set the `nonce` attribute for all scripts/style tags
// i.e will run styleTag.setAttribute('nonce', 'r4nd0m')
// Obviously, you should use the nonce generated by your server
setNonce('r4nd0m');

// Generic Angular bootstrapping
platformBrowserDynamic()
  .bootstrapModule(AppModule)
  .catch(err => console.log(err));
```

#### WWW

Unfortunately, setting `nonce` attributes gets a bit trickier when it comes to [SSR and SSG](docs/static-site-generation). As a `nonce` needs
to be unique per page view, we cannot define/set the nonce at build time. So, this responsibility now falls on the

> [hydrate app](docs/hydrate-app)'s execution of runtime code.

Reviewer: IIRC, these links need a leading slash. ATM, this points to docs/docs/hydrate-app in the generated docs

```suggestion
[hydrate app](/docs/hydrate-app)'s execution of runtime code.
```

> ##### SSR

Reviewer: I don't think the Stencil Site supports 5 pound signs at the moment:
![Screenshot 2022-12-15 at 3 37 26 PM](https://user-images.githubusercontent.com/1930213/207961994-412a568e-bf71-4355-99d7-4d61ddda47ab.png)

Can we reorder/restructure this doc to work around this limitation?

Reviewer 2: does this still need to get updated?

Author: Nope, sorry, forgot to mark this resolved. I replaced those headers with bolded elements. Maybe we can swap this back if the new docs site supports up to h5/h6

Since there is not an easy way (or any way) of exposing and executing helper functions to manipulate the outcome of the runtime code, Stencil
has fallback behavior for pulling the `nonce` off of the window object.

So, for SSR, your app can simply set the desired `nonce` value on `window.nonce` at bootstrap. Yes, this does involve some manual configuration

> for the code served by your server.

Reviewer: Can we update this to clarify that 'bootstrap' should be when the page is rendered? I'd like to make it clear that this shouldn't just happen once, but every time a server serves the page.

// main.ts

import { defineCustomElements, setNonce } from 'my-lib/dist/components';

> // Assume `autoDefineCustomElements` is enabled enabled

Reviewer: How do we want to handle this once `autoDefineCustomElements` goes away in v3.0.0?

Author: Once this lands and we rebase v3, I can put up a small PR to change this comment to the export behavior instead
